File: Android App/app/src/main/java/com/fica/events/ui/components/SponsorImage.kt
package com.fica.events.ui.components

import com.fica.events.R

/**
 * Maps a sponsor's `logo_url` to a bundled drawable resource id when we
 * ship the PNG inside the APK. Returns `null` for any URL we don't have
 * packaged — in that case the caller should fall back to Coil AsyncImage
 * on the URL.
 *
 * The sponsor roster is locked for FICA Congress 2026 — shipping the
 * logos inside the app means they render instantly on first paint, no
 * flicker, no network dependency.
 */
object SponsorImage {

    // logo_url filename → drawable res id.
    private val mapping: Map<String, Int> = mapOf(
        "vodafone.png"         to R.drawable.sponsor_vodafone,
        "asco-motors.png"      to R.drawable.sponsor_asco_motors,
        "extra.png"            to R.drawable.sponsor_extra,
        "merchant-finance.png" to R.drawable.sponsor_merchant_finance,
        "marsh.png"            to R.drawable.sponsor_marsh,
        "motibhai.png"         to R.drawable.sponsor_motibhai,
        "ca-anz.png"           to R.drawable.sponsor_ca_anz,
        "datec.png"            to R.drawable.sponsor_datec,
        "fiji-airways.png"     to R.drawable.sponsor_fiji_airways,
        "fiji-times.png"       to R.drawable.sponsor_fiji_times,
    )

    fun bundledResFor(logoUrl: String?): Int? {
        val raw = logoUrl?.takeIf { it.isNotBlank() } ?: return null
        // Strip any query string, pull the last path segment, lowercase.
        val filename = raw
            .substringBefore('?')
            .substringBefore('#')
            .substringAfterLast('/')
            .lowercase()
        return mapping[filename]
    }
}

File: IOS App/Events/Services/SponsorImage.swift
import Foundation

/// Maps a sponsor's `logo_url` to a bundled asset catalog name when the
/// sponsor is one of the known pre-packaged ones. Returning `nil` means
/// "I don't have this one bundled — fall back to network."
///
/// The sponsor roster is locked for FICA Congress 2026 — shipping the
/// PNGs inside the app means they render instantly on first launch,
/// survive a bad network, and never flash a placeholder.
enum SponsorImage {
    /// Filename-component → asset catalog name.
    private static let mapping: [String: String] = [
        "vodafone.png":         "SponsorVodafone",
        "asco-motors.png":      "SponsorAscoMotors",
        "extra.png":            "SponsorExtra",
        "merchant-finance.png": "SponsorMerchantFinance",
        "marsh.png":            "SponsorMarsh",
        "motibhai.png":         "SponsorMotibhai",
        "ca-anz.png":           "SponsorCaAnz",
        "datec.png":            "SponsorDatec",
        "fiji-airways.png":     "SponsorFijiAirways",
        "fiji-times.png":       "SponsorFijiTimes",
    ]

    /// Returns the asset catalog name for a bundled sponsor logo matching
    /// the given URL, or nil if the URL isn't one of the known sponsors.
    static func bundledAsset(for urlString: String?) -> String? {
        guard let raw = urlString, !raw.isEmpty else { return nil }
        // Lowercase the last path component so we tolerate case drift or
        // query strings (e.g. `?v=2`); mapping keys are lowercase.
        let filename: String
        if let url = URL(string: raw) {
            filename = url.lastPathComponent.lowercased()
        } else {
            // Raw string with no valid URL structure — try splitting.
            filename = raw.split(separator: "/").last.map(String.init)?.lowercased() ?? ""
        }
        return mapping[filename]
    }
}
